grype/db/v6/models.go

```diff
@@ -275,15 +282,53 @@ func (p *Package) BeforeCreate(tx *gorm.DB) (err error) {
 	return nil
 }
 
+// OperatingSystem represents specific release of an operating system. The resolution of the version is
+// relative to the available data by the vulnerability data provider, so though there may be major.minor.patch OS
+// releases, there may only be data available for major.minor.
 type OperatingSystem struct {
 	ID ID `gorm:"column:id;primaryKey"`
 
-	Name         string `gorm:"column:name;index:os_idx,unique;index"`
-	ReleaseID    string `gorm:"column:release_id;index:os_idx,unique;index"`
+	// Name is the operating system family name (e.g. "debian")
+	Name      string `gorm:"column:name;index:os_idx,unique;index"`
+	ReleaseID string `gorm:"column:release_id;index:os_idx,unique;index"`
+
+	// MajorVersion is the major version of a specific release (e.g. "10" for debian 10)
 	MajorVersion string `gorm:"column:major_version;index:os_idx,unique;index"`
+
+	// MinorVersion is the minor version of a specific release (e.g. "1" for debian 10.1)
 	MinorVersion string `gorm:"column:minor_version;index:os_idx,unique;index"`
+
+	// LabelVersion is an optional non-codename string representation of the version (e.g. "unstable" or for debian:sid)
 	LabelVersion string `gorm:"column:label_version;index:os_idx,unique;index"`
-	Codename     string `gorm:"column:codename;index"`
+
+	// Codename is the codename of a specific release (e.g. "buster" for debian 10)
+	Codename string `gorm:"column:codename;index"`
+}
+
+func (os *OperatingSystem) VersionNumber() string {
+	if os.MinorVersion != "" {
+		return fmt.Sprintf("%s.%s", os.MajorVersion, os.MinorVersion)
+	}
+	return os.MajorVersion
+}
+
+func (os *OperatingSystem) Version() string {
```

Comment:
Not a blocking comment, but I'm curious: What motivates this particular fallback logic?

Reply:
Here's the reasoning that went into field precedence:
- label version tends to override numeric versions ("unstable" is the label for "debian:sid" for instance). There is a good argument that codename should be here, but the only reason why it's not is because vunnel and grype-db reason about a value here over raw values from the vulnerability provider.
- if not found, then we craft the most specific version possible and return that
- we don't expect there to be many cases where there is no version information but if so, codename is the only recourse left (which tends to be empty though).